core/lib/Cardano/Address/Style/Shelley.hs

```diff
@@ -683,48 +714,63 @@ data ErrExtendAddress
 
 -- Re-export from 'Cardano.Address' to have it documented specialized in Haddock.
 --
--- | Convert a public key and a staking key to a delegation 'Address' valid
--- for the given network discrimination. Funds sent to this address will be
--- delegated according to the delegation settings attached to the delegation
--- key.
+-- | Convert a payment credential (key or script) and a staking credential (key or script)
+-- to a delegation 'Address' valid for the given network discrimination.
+-- Funds sent to this address will be delegated according to the delegation settings
+-- attached to the delegation key.
 --
 -- @since 2.0.0
 delegationAddress
     :: NetworkDiscriminant Shelley
-    -> Shelley 'AddressK XPub
-    -> Shelley 'StakingK XPub
+    -> PaymentCredential
+    -> StakeCredential
     -> Address
-delegationAddress =
-    Internal.delegationAddress
+delegationAddress discrimination paymentCredential stakeCredential =
+    unsafeFromRight $ extendAddress
+        (paymentAddress discrimination paymentCredential)
+        (Left stakeCredential)
 
 
 -- Re-export from 'Cardano.Address' to have it documented specialized in Haddock.
 --
--- | Convert a public key and pointer to staking certificate in blockchain to a
+-- | Convert a payment credential (key or script) and pointer to staking certificate in blockchain to a
 -- pointer 'Address' valid for the given network discrimination.
 --
 -- @since 2.0.0
 pointerAddress
     :: NetworkDiscriminant Shelley
-    -> Shelley 'AddressK XPub
+    -> PaymentCredential
     -> ChainPointer
     -> Address
-pointerAddress =
-    Internal.pointerAddress
+pointerAddress discrimination credential pointer =
+    case credential of
+        PaymentFromKey keyPub ->
+            -- pointer address: keyhash28, 3 variable length uint : 01000000 -> 64
+            Internal.pointerAddress discrimination keyPub pointer
+        PaymentFromScript scriptHash ->
+           -- pointer address: scripthash32, 3 variable length uint : 01010000 -> 80
+            unsafeFromRight $ extendAddress
+            (paymentAddress discrimination (PaymentFromScript scriptHash))
+            (Right pointer)
```

Comment:
Why aren't credentials passed to `pointerAddress` directly ? (and same question for `paymentAddress` actually).. 
Wouldn't be easier to modify the signature of the classes above or, to not have the type-classes at actually. 

I mean, in the end, the type-classes would be nice if we wanted to treat credentials and addresses using one single interface but in the end, application code often treats Byron addresses differently from Shelley addresses _anyway_. So there's not much value in capturing this in classes at all.

Reply:
agree Pointer/Delegation/Payment impl are redundant here as they capture only subset of arguments -> removed them, needed to relocate some testing from Cardano.AddressesSpec to ShelleySpec as a consequence. Done in https://github.com/input-output-hk/cardano-addresses/pull/79/commits/c2df3de55ba2bcdbba949cfd6468a4372a9b1751